### PR TITLE
MM-46410: adds urgency on mention counts

### PR DESCRIPTION
**Pull request by koox00**
#### Summary

We have introduced priority for posts in mattermost/mattermost-webapp#10951. 
We do need, now, to color the mention badges in the webapp with a prominent color when a mention is posted in an urgent message. 

In the case of threads, all mentions in a thread with a root post marked as urgent,
are considered urgent mentions.

This PR adds one column, `urgentmentioncount`, in channelmembers table.
Furthermore when asking for team/thread mention counts, we also return urgent mention counts for the user.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-46410

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by koox00** [comment]
FYI, I'll be adding unit tests shortly.

**Comment by lieut-data** [comment]
@koox00, at a high level, how many threads are we expecting to be marked as "urgent"? Before we consider adding new columns, I'd love to verify that these won't just be sparsely used (but still have the performance impact), otherwise we might consider a different approach for tracking urgent posts.

Is there a design document for this feature I can reference?

**Comment by agnivade** [comment]
@koox00 - Before we go into the review, I'd like to see the results of a load test with the feature flag enabled. Would you be able to give that a spin?

**Comment by larkox** [comment]
This changes a model that is in the graphql schema. Consider updating `api4/schema.graphqls` and any other relevant file to add this to the schema.

On front-end implementations of this, keep in mind potential changes needed in the graphql branch of things.

Here is an example PR on how to add a field to the schema, but if you have any problem, feel free to ask:
https://github.com/mattermost/mattermost-server/pull/20956

**Comment by enahum** [comment]
Added a comment in https://github.com/mattermost/mattermost-webapp/pull/11116#issuecomment-1245424607 to see if we have given it any consideration

File: api4/schema.graphqls
Comment (agnivade) [inline api4/schema.graphqls:79]: Lovely to see the gql schema extended. Please also add a test for the existence of the field in api4/resolver_channel_member_test.go. Thanks!

**Comment by koox00** [comment]
> @koox00 - Before we go into the review, I'd like to see the results of a load test with the feature flag enabled. Would you be able to give that a spin?

@agnivade sorry for taking so long, here's the grafana for the load test http://ec2-3-88-39-35.compute-1.amazonaws.com:3000/?orgId=1&from=1666359000000&to=1666363200000&var-server=app-0:8067&var-server=app-0:9100

do you prefer me to post screenshots?

**Comment by agnivade** [comment]
@agarciamontoro - could you check the loadtest results?

@koox00 - Do we need to add any extra coverage in the loadtest to ensure the new code paths are executed? Or they will always run in the normal flow?

**Comment by koox00** [comment]
> @koox00 - Do we need to add any extra coverage in the loadtest to ensure the new code paths are executed? Or they will always run in the normal flow?

@agnivade they will always run in the normal flow. 
when a channel/thread is viewed or marked as unread, when a new post/thread is created, and when we request for teams unread counts.

**Comment by agnivade** [comment]
Thanks, I will leave it to Alejandro to give the results a check.

**Comment by agarciamontoro** [comment]
Thanks for the ping, @agnivade.

@koox00, are these unbounded tests? If so, it looks like it was capped by a maximum number of 4K users. Can I check the configuration files you used to understand the setup?

**Comment by agarciamontoro** [comment]
Quick update: @koox00 and I synced and tests are being repeated to remove the artificial 4k cap we saw there.

File: app/post.go
Comment (koox00) [inline app/post.go:282]: @lieut-data we need to reinstate the metadata, since that's the field we are using to save post priority values.
Do you see any reason that this could cause problems?
Do plugins set their own `metadata`, would we need to merge `PostMetadata.Priority`?
cc @lieut-data 

BTW, `post.ForPlugin` clones the post, and then nils the post.Metadata for the clone.
Since cloning is a shallow copy, wouldn't that make the original post's metadata nil?
I wonder if we mind to deep copy metadata in the `post.Clone` method

**Comment by koox00** [comment]
@agnivade load test results added.

avg-api-times
avg-store-times
cpu-utilization
goroutines-in-use
heap-in-use
p99-api-times
p99-store-times
rps
stack-in-use
[results.txt](https://github.com/mattermost/mattermost-server/files/9943834/results.txt)


PS: I've made some changes to the load test to create 5% of posts as urgent posts in the simulcontroller, and (by mistake) 50% urgent posts in the gencontroller

File: app/post.go
Comment (lieut-data) [reply app/post.go:282]: Hey @koox00! There's potentially a couple of layers to this question, but it all sounds like we need to rethink what it means to allow a plugin to replace post as part of processing the `MessageWillBePosted`. Are some fields off limits, and the server will forcibly rebuild them? Or can plugins act unilaterally in this regard?

cc @catalintomai / @levb

**Comment by koox00** [comment]
@agnivade load test results added.

avg-api-times
avg-store-times
cpu-utilization
goroutines-in-use
heap-in-use
p99-api-times
p99-store-times
rps
stack-in-use
[results.txt](https://github.com/mattermost/mattermost-server/files/9944635/results.txt)

PS: I've made some changes to the load test to create 5% of posts as urgent posts in the simulcontroller, and (by mistake) 50% urgent posts in the gencontroller

**Comment by lieut-data** [comment]
@koox00, thanks for the effort in setting up and running loadtests! Something seems off about the results, in that -- at first glance, at least -- adding this feature dramatically improves performance across the board. 

From my review of the code, I'm not sure there's a reason we should expect such results. Is there any extra context you can add about the loadtests? @agarciamontoro, any thoughts as to the interpretation here?

**Comment by koox00** [comment]
> @koox00, thanks for the effort in setting up and running loadtests! Something seems off about the results, in that -- at first glance, at least -- adding this feature dramatically improves performance across the board.
> 
> From my review of the code, I'm not sure there's a reason we should expect such results. Is there any extra context you can add about the loadtests? @agarciamontoro, any thoughts as to the interpretation here?

@lieut-data writing tests I realized I had a syntax error in code, don't know how and if it affected the results but I running the tests again.

**Comment by agarciamontoro** [comment]
I also think there's something weird going on there: either the variability of tests or some other error, so thank you for repeating the tests, @koox00. Maybe a comparison between bounded tests with the same number of users for a long period of time would be more insightful.

It's also interesting to note that the graphs show a decrease in goroutines, stack, heap and CPU usage, but I don't see any significant improvement in the store and API times (maybe with the flag on they have less spikes, but the increase/decrease of users in the unbounded test may be playing a role there). Also, the number of requests per second seems to be lower when the flag is on.

File: app/post.go
Comment (levb) [reply app/post.go:282]: The most outward "layer" is that `Metadata` was a rather extensible struct, containing 3rd party types. When these types change, we end up causing backwards compatibility issues with previously compiled plugins, leading to crashes.

One solution would be to separate the "plugin-visible" metadata into another struct, and maintain the type version discipline on the new struct. We wold **still** have issues with older plugins, so this would be a "new API" change?

FWIW, I am not sure that restoring the old posts' metadata to the modified post is the right thing to do.

**Comment by koox00** [comment]
@lieut-data, @agarciamontoro these are the new results, seems like the same as before

avg-store-times
cpu-utilization
goroutines-in-use
heap-in-use
p99-store-times
rps
stack-in-use
[results.txt](https://github.com/mattermost/mattermost-server/files/9962637/results.txt)

File: app/post.go
Comment (koox00) [reply app/post.go:282]: > FWIW, I am not sure that restoring the old posts' metadata to the modified post is the right thing to do.

@levb would restoring the `metadata.priority` make sense? Since that's what we want to handle through the app and not a plugin.

**Comment by agarciamontoro** [comment]
@koox00, this is still quite odd, and mildly concerning, since the RPS do show a significant decrease. Is the Grafana for these deployments still up?

**Comment by koox00** [comment]
> @koox00, this is still quite odd, and mildly concerning, since the RPS do show a significant decrease. Is the Grafana for these deployments still up?

The one with the feature flag off is still up http://ec2-54-83-70-144.compute-1.amazonaws.com:3000/?orgId=1&from=1667914740000&to=1667918340000&var-server=app-0:8067

**Comment by koox00** [comment]
@lieut-data, @agarciamontoro these are the changes to the load test I'm using to test this PR

https://github.com/mattermost/mattermost-load-test-ng/compare/master...koox00:MM-46410_priority-badges?expand=1

**Comment by agarciamontoro** [comment]
Changes look good, but one thing caught my attention. What's the new last parameter to `GetPosts...` functions? Can that affect the test?
```go
postList, _, err := ue.client.GetPostsForChannel(channelId, page, perPage, "", collapsedThreads, false)
postList, _, err := ue.client.GetPostsBefore(channelId, postId, page, perPage, "", collapsedThreads, false)
postList, _, err := ue.client.GetPostsAfter(channelId, postId, page, perPage, "", collapsedThreads, false)
```

**Comment by agarciamontoro** [comment]
The only thing we haven't tried is repeating these tests with bounded tests, instead of unbounded. Those should be more stable, and maybe they get rid of the weird behaviour we're seeing.

**Comment by koox00** [comment]
I've setup 2 bounded load tests, one with the flag on, and one with the flag off, 
using the 12M posts db, 
in which for the 5% of posts, in a random order, I've created rows in `PostsPriority` table, 
so 600K "urgent" posts in both DBs.

Dashboards

- priority enabled: http://ec2-184-73-148-142.compute-1.amazonaws.com:3000/?orgId=1&refresh=10s&from=now-30m&to=now&var-server=app-0:8067
- priority disabled: http://ec2-54-209-84-53.compute-1.amazonaws.com:3000/?orgId=1&refresh=10s&var-server=app-0:8067&from=now-30m&to=now

I hope this will give clearer results.

**Comment by agnivade** [comment]
(As a side note)

@koox00 - Is there a PR for your adding the message priority label to the load tests? We should probably add that to the codebase.

**Comment by koox00** [comment]
> @koox00 - Is there a PR for your adding the message priority label to the load tests? We should probably add that to the codebase.

@agnivade I have the branch if you want to take a quick look. I'll open a PR now as well!

**Comment by lieut-data** [comment]
Thanks for posting the direct Grafana links, @koox00! I recall @agnivade pointing something out in some of my past load tests that when I see a metric flatlining at 10s, it translates into a timeout and makes it hard to properly interpret the results. Installing the [Mattermost Performance Monitoring v2](https://grafana.com/grafana/dashboards/15582-mattermost-performance-monitoring-v2/) dashboard, I see this for /both/ clusters on the create post and channel load duration measurements:

CleanShot 2022-11-10 at 11 45 02@2x
CleanShot 2022-11-10 at 11 47 15@2x

A theory: perhaps we scale up the database instance to properly handle the load, and then we can compare the respective performance?

@agarciamontoro, is there an opportunity to highlight these timeouts in the graphs generated above so we can immediately identify an under-resourced loadtest?

**Comment by agarciamontoro** [comment]
> @agarciamontoro, is there an opportunity to highlight these timeouts in the graphs generated above so we can immediately identify an under-resourced loadtest?

@lieut-data, Do you mean in the comparison graphs?

**Comment by lieut-data** [comment]
> @lieut-data, Do you mean in the comparison graphs?

Yeah, I'm wondering if there's a way to surface that "something is broken." Maybe it's as simple as adding a graph for the create post and channel load metrics we already expose in Grafana? Just to help surface this kind of issue more quickly for subsequent feature loadtests.

**Comment by koox00** [comment]
@lieut-data, @agarciamontoro new deployments, with 2 app-instances

Dashboards

    priority enabled: http://ec2-44-201-140-114.compute-1.amazonaws.com:3000/?orgId=1&refresh=10s&var-server=app-1:8067&var-server=app-1:9100&from=now-30m&to=now
    priority disabled: http://ec2-44-206-230-7.compute-1.amazonaws.com:3000/?orgId=1&refresh=10s&var-server=app-1:8067&var-server=app-0:8067&from=now-30m&to=now

**Comment by agarciamontoro** [comment]
That looks much better! Looking forward to see the final comparison :)

**Comment by lieut-data** [comment]
🎉 -- awesome to have been able to sort out the issues here. Great persistence, @koox00, and thanks for the excellent support @agarciamontoro :)

**Comment by ashishbhate** [comment]
Following this discussion on community and here - I'm not sure I understand why the store calls were timing out? If having only one app server (as opposed to the usual two) was the bottleneck, then why where the DB calls timing out? 
Theoretically the load on the DB was half of the expected load (one app server vs two app servers)

File: app/post_metadata.go
Comment (lieut-data) [inline app/post_metadata.go:134]: Would I be correct in tracing the code to see `App.PreparePostListForClient` effectively calling this method in a loop for a set of posts, e.g. when fetching root posts in a channel? Do we need a means to fetch this priority in bulk instead?

File: store/sqlstore/channel_store.go
Comment (lieut-data) [inline store/sqlstore/channel_store.go:2557]:
```suggestion
		query = query.Where(sq.Eq{"Posts.UserId": userId})
```

File: store/sqlstore/channel_store.go
Comment (lieut-data) [inline store/sqlstore/channel_store.go:2561]: nit: let's include this `Where` condition above, and have a dedicated line for `query.ToSql`.

File: store/sqlstore/channel_store.go
Comment (lieut-data) [inline store/sqlstore/channel_store.go:2547]: It shouldn't matter from the database perspective since it ought to optimize to the same result, but would it make sense to write this as a query `FROM PostsPriority` and `JOIN Posts` to clarify the source of truth at play?

File: store/sqlstore/post_priority_store.go
Comment (lieut-data) [inline store/sqlstore/post_priority_store.go:27]: Could we use `GetBuilder` to avoid having to manually extract the args?

File: store/sqlstore/post_priority_store.go
Comment (lieut-data) [inline store/sqlstore/post_priority_store.go:34]: nit:
```suggestion
```

File: store/sqlstore/post_priority_store.go
Comment (lieut-data) [inline store/sqlstore/post_priority_store.go:36]: nit:
```suggestion
```

File: store/sqlstore/thread_store.go
Comment (lieut-data) [inline store/sqlstore/thread_store.go:227]: Should this be `JOIN` given we require a `Priority` match in the conditions below?

File: store/sqlstore/thread_store.go
Comment (lieut-data) [inline store/sqlstore/thread_store.go:226]: Should we make this join conditional within the `teamId != ""` check below to be explicit? (Again, database will ignore this, but I find it helpful to clarify the happy path for SQL when reading.)

File: store/sqlstore/thread_store.go
Comment (lieut-data) [reply store/sqlstore/thread_store.go:286]: Also, can we group all the affected lines together?

```go
if (featureFlagEnabled) {
    urgencyCase := ...
    query = query.
        Column(...)
        LeftJoin(...)
}
```

File: store/sqlstore/thread_store.go
Comment (lieut-data) [inline store/sqlstore/thread_store.go:380]: Should we consider using https://pkg.go.dev/golang.org/x/sync/errgroup?

File: store/sqlstore/thread_store.go
Comment (lieut-data) [inline store/sqlstore/thread_store.go:439]: Should this be a non-left `JOIN` to be explicit?

**Review by lieut-data — CHANGES_REQUESTED** [review]
Thanks for walking me through some of this last week, @koox00! A few comments below. 

Also, do we need to consider the import/export case for this metadata?

**Comment by agarciamontoro** [comment]
The average and P99 API time graphs are a bit concerning, they are consistently slower with the feature flag on, which also matches the increase in CPU utilization. However, there are not significant differences in store times and RPS, which is interesting. @koox00, any idea why the changes here would affect CPU utilization and API times that much?

File: app/post_metadata.go
Comment (koox00) [reply app/post_metadata.go]: I've made a try on this, so when requesting PreparePostListForClient priority metadata are fetched in bulk, but we also need to, optionally, add priority when we prepare a single post for the client (PreparePostForClient)

**Comment by agnivade** [comment]
I am removing myself. Alejandro's review on the load tests are sufficient :)

File: db/migrations/postgres/000097_create_posts_priority.up.sql
Comment (ashishbhate) [inline db/migrations/postgres/000097_create_posts_priority.up.sql:4]: Why 32?

**Review by ashishbhate — COMMENTED** [review]
Great job @koox00 did a first pass review with a few comments. I'll do another pass soon.

File: db/migrations/postgres/000097_create_posts_priority.up.sql
Comment (koox00) [reply db/migrations/postgres/000097_create_posts_priority.up.sql:4]: I think there is a plan to support custom labels in the future. I thought 32 would be a good number, but it is indeed random.

**Comment by agarciamontoro** [comment]
Regarding the latest results in https://github.com/mattermost/mattermost-server/pull/20999#issuecomment-1316717958:

Those API times do look better! That's less of a concern for me now, although I still don't understand (as you pointed out) why the CPU usage is that high.

For sharing it publicly: Kyriakos and I discussed privately that the next tests after the improvement he's doing should be done in the same instance, to remove the variability that comes from two different instances (even if they're of the same class). Let's see if that helps to get rid of the CPU differences.

File: store/sqlstore/thread_store.go
Comment (lieut-data) [reply store/sqlstore/thread_store.go]: `ThreadDeleteAt` is a new column without a default as of v7.6. While we migrate it on upgrade, in theory I suppose a concurrent write to the table from an old version of the software might add a `null` record.

File: store/sqlstore/thread_store.go
Comment (lieut-data) [reply store/sqlstore/thread_store.go]: counter nit: that's cool behaviour for `errors.Wrap`, but I feel like I'd stumble over reading that compared to the more idiomatic `return nil` after a successful execution. 0/5

File: app/post_priority.go
Comment (lieut-data) [inline app/post_priority.go:31]: Is there any particular reason to batch at the app layer vs. just passing all the ids to the store? (Seems like the database is in the best position to deal with the batching if any is required, plus this code is simpler.)

**Review by lieut-data — APPROVED** [review]
This is really terrific work, @koox00 -- amazing follow up to all the conversation on this PR. You're really "owning this feature" :)

Two non-blocking comments below.

File: app/post_priority.go
Comment (koox00) [reply app/post_priority.go:31]: oh, no reason at all, I saw that we were doing batching in the app layer in app/post, I think, that's why I went this route.
I'll change it as it indeed makes more sense

**Comment by koox00** [comment]
@agarciamontoro I've run the tests in the same server, first the on and then the off,
attaching the comparison results below. grafana [link](http://ec2-35-174-171-73.compute-1.amazonaws.com:3000/?orgId=1&from=1668615581079&to=1668631445194&var-server=agent-0:4000&var-server=agent-0:9100&var-server=agent-1:4000&var-server=agent-1:9100&var-server=agent-2:4000&var-server=agent-2:9100&var-server=agent-3:4000&var-server=agent-3:9100&var-server=app-0:8067&var-server=app-0:9100&var-server=app-1:8067&var-server=app-1:9100&var-server=localhost:9090&var-server=localhost:9100&var-server=metrics:9090&var-server=metrics:9100&var-server=proxy:9100)
avg-api-times
avg-store-times
cpu-utilization
goroutines-in-use
heap-in-use
number-of-connected-devices
p99-api-times
p99-store-times
rps
stack-in-use
[results.txt](https://github.com/mattermost/mattermost-server/files/10025555/results.txt)

File: api4/user_test.go
Comment (ashishbhate) [inline api4/user_test.go:6590]: Why don't we use the constant defined in the model?

**Review by ashishbhate — APPROVED** [review]
fantastic stuff @koox00 !

**Comment by agarciamontoro** [comment]
@koox00, those graphs look now much better. No performance hit overall, as far as I can see. The only thing that catches the eye is the heap in use graph, but those numbers fluctuate even for the off case among the different tests you've done, so I'm comfortable discarding that as measurement variability and wouldn't spend more time on this.

I'm curious to know if you had a chance to query and compare the specific API endpoints affected by the PR. That'd be interesting to see.

In any case, from a performance point of view, this looks good now. Thank you for your infinite patience on this, I'll schedule a retrospective with you so we can summarize the main pain points faced during all this.

**Comment by jgilliam17** [comment]
Thanks @koox00 
I am still having intermittent issues with marking posts as unread, any post, not just ones with post priority label.
1. I found it confusing when marking an older, Urgent with mention, post as unread and having all other bellow mentions in the channel included in the count. Made  it seem that I have 6 urgent messages/mentions, while the true count is 2.
Screen Shot 2022-11-17 at 4 07 21 PM
2. When marking the post as unread, sometimes coloring of the badge is lagging. Takes a bit to turn red.
Screen Shot 2022-11-17 at 4 07 58 PM

**Comment by koox00** [comment]
> I found it confusing when marking an older, Urgent with mention, post as unread and having all other bellow mentions in the channel included in the count. Made it seem that I have 6 urgent messages/mentions, while the true count is 2.

@jgilliam17 I see.. The way to think about it is that you indeed have 6 unread mentions, and at least 1 of them is "urgent".

>  2. When marking the post as unread, sometimes coloring of the badge is lagging. Takes a bit to turn red. 

Is it OK to add a follow up ticket about this?

**Comment by Willyfrog** [comment]
Regarding 1, I believe it is by design, maybe something we can iterate later on cc/ @matthewbirtch 

> Is it OK to add a follow up ticket about this?

yes, please create one

**Comment by matthewbirtch** [comment]
> Regarding 1, I believe it is by design, maybe something we can iterate later on cc/ @matthewbirtch 

Not actually by design. I don't know if this specific case was thought about. I definitely see the problem/confusion and I think the only true way to solve this would be to have two mention counts side by side, but that's also very problematic.

**Comment by koox00** [comment]
> Not actually by design. I don't know if this specific case was thought about. I definitely see the problem/confusion and I think the only true way to solve this would be to have two mention counts side by side, but that's also very problematic. I don't think I have a decent solution off the top of my head right now.

I had the impression as well that it was by design.

**Comment by esethna** [comment]
@koox00 @Rishselva I'm going to bump this PR to v7.7 so we can narrow our focus to the ack and persistent notifications PRs for the feature complete deadline. The mention badge coloring is a nice to have, not a feature priority.

**Comment by koox00** [comment]
> @koox00 @Rishselva I'm going to bump this PR to v7.7 so we can narrow our focus to the ack and persistent notifications PRs for the feature complete deadline. The mention badge coloring is a nice to have, not a feature priority.

@esethna this PR is required for both ack, and notifications to be merged

**Comment by esethna** [comment]
Ah, thanks for clarifying Kyriakos

**Comment by esethna** [comment]
@jgilliam17 can we approve this and open a new ticket for the second issue? Please merge when that's done, thanks!

**Comment by koox00** [comment]
@jgilliam17, @esethna please don't merge, I'll merge when Jelena approves :)

**Comment by jgilliam17** [comment]
Thanks @koox00 
Tested, looks good to merge.
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/263923-df534ef-onprem-ent-server-pr-20999-mm-ee-test:df534ef/mochawesome.html) - few failures with accessibility in settings/profile modals, doesn't appear elated to changes here
Ticket filed for follow up, delay in badge turning red https://mattermost.atlassian.net/browse/MM-48565

**Review by jgilliam17 — APPROVED** [review]

**Comment by jgilliam17** [comment]
@koox00 E2E report looks good - https://mattermost-cypress-report.s3.amazonaws.com/265875-875dbc7-onprem-ent-server-pr-20999-mm-ee-test:875dbc7/mochawesome.html same failures as master

File: app/post.go
Comment (wiggin77) [inline app/post.go:1840]: @koox00 did you intend to check `nErr` for not nil here?

Comment (wiggin77) [reply app/post.go:1840]: Also, I wonder how the linter missed that.

Comment (agnivade) [reply app/post.go:1840]: Oh wow, great catch! Yeah the linter sometimes misses things like this.

**Pull request merged**
